test/python/circuit/test_hamiltonian_gate.py

```diff
@@ -0,0 +1,201 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=arguments-differ,method-hidden
+
+"""Quick program to test the qi tools modules."""
+
+import json
+import numpy as np
+from numpy.testing import assert_allclose
+
+
+import qiskit
+from qiskit.extensions.hamiltonian_gate import HamiltonianGate, UnitaryGate
+from qiskit.extensions.exceptions import ExtensionError
+from qiskit.test import QiskitTestCase
+from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
+from qiskit.circuit import Parameter
+from qiskit.quantum_info import Operator
+from qiskit.transpiler import PassManager
+from qiskit.compiler import transpile
+from qiskit.converters import circuit_to_dag, dag_to_circuit
+from qiskit.transpiler.passes import CXCancellation
+
+
+class TestHamiltonianGate(QiskitTestCase):
+    """Tests for the HamiltonianGate class."""
+
+    def test_set_matrix(self):
+        """Test instantiation"""
+        hamiltonian = HamiltonianGate([[0, 1], [1, 0]], 1)
+        self.assertEqual(hamiltonian.num_qubits, 1)
+
+    def test_set_matrix_raises(self):
+        """test non-unitary"""
+        with self.assertRaises(ExtensionError):
+            HamiltonianGate([[1, 0], [1, 1]], 1)
+
+    def test_conjugate(self):
+        """test conjugate"""
+        ham = HamiltonianGate([[0, 1j], [-1j, 2]], np.pi / 4)
+        np.testing.assert_array_almost_equal(ham.conjugate().to_matrix(),
+                                             np.conj(ham.to_matrix()))
+
+    def test_transpose(self):
+        """test transpose"""
+        ham = HamiltonianGate([[15, 1j], [-1j, -2]], np.pi / 7)
+        np.testing.assert_array_almost_equal(ham.transpose().to_matrix(),
+                                             np.transpose(ham.to_matrix()))
+
+    def test_adjoint(self):
+        """test adjoint operation"""
+        ham = HamiltonianGate([[3, 4j], [-4j, -.2]], np.pi * .143)
+        np.testing.assert_array_almost_equal(ham.adjoint().to_matrix(),
+                                             np.transpose(np.conj(ham.to_matrix())))
+
+
+class TestHamiltonianCircuit(QiskitTestCase):
+    """Hamiltonian gate circuit tests."""
+
+    def test_1q_hamiltonian(self):
+        """test 1 qubit hamiltonian"""
+        qr = QuantumRegister(1, 'q0')
+        cr = ClassicalRegister(1, 'c0')
+        qc = QuantumCircuit(qr, cr)
+        matrix = np.zeros((2, 2))
+        qc.x(qr[0])
+        theta = Parameter('theta')
+        qc.append(HamiltonianGate(matrix, theta), [qr[0]])
+        qc = qc.bind_parameters({theta: 1})
+
+        # test of qasm output
+        self.log.info(qc.qasm())
+        ref_qasm = "OPENQASM 2.0;\n" \
+                   "include \"qelib1.inc\";\n" \
+                   "qreg q0[1];\n" \
+                   "creg c0[1];\n" \
+                   "x q0[0];\n" \
+                   "hamiltonian([[0.+0.j 0.+0.j]\n" \
```

Comment:
Strictly speaking, this is not OpenQASM 2.0. I'm not sure if this can be solved like in UnitaryGate, tho...

Reply:
I don't think it can. Should I just take out the ability to generate QASM and throw an error if the user tries?

Reply:
sounds good to me.. maybe @ajavadia or @kdk have an opinion here.

Reply:
Okay with raising an error on `qasm()` (and a follow up issue to try to find a better way). For `UnitaryGate`, we synthesize on call to `qasm` (because QASM doesn't support array-like parameters). Maybe a similar hack could work here.

Reply:
Ok, done.